### PR TITLE
Use futuresordered in join_all

### LINKED ISSUE #2201: Shall quadratic complexity of `join_all` be documented?
* https://github.com/tokio-rs/tokio/issues/2401
* https://users.rust-lang.org/t/hyper-tonic-client-performance/47253/2

### PR DESCRIPTION
**Pull request by ibraheemdev**
Switch to `FuturesOrdered` in `join_all` when dealing with large amounts of futures. We will probably need benchmarks to determine what `SMALL` is.

Closes #2201

### CODE REVIEW

**Comment by ibraheemdev** [comment]
I think CI is failing because `FuturesUnordered` is feature gated under `cfg_target_has_atomic`?

**Comment by taiki-e** [comment]
> I think CI is failing because `FuturesUnordered` is feature gated under `cfg_target_has_atomic`?

Yeah, I think you probably have to use `#[cfg(not(futures_no_atomic_cas))]` on `JoinAllKind::Big`. (In other words, if atomic CAS is not available, always use `JoinAllKind::Small`.)

**Comment by taiki-e** [comment]
> > I think CI is failing because `FuturesUnordered` is feature gated under `cfg_target_has_atomic`?
> 
> Yeah, I think you probably have to use `#[cfg(not(futures_no_atomic_cas))]` on `JoinAllKind::Big`. (In other words, if atomic CAS is not available, always use `JoinAllKind::Small`.)

Btw, at the time this PR was opened, it was cfg_target_has_atomic, but now it is futures_no_atomic_cas after https://github.com/rust-lang/futures-rs/pull/2400 has been merged.

**Comment by ibraheemdev** [comment]
```
 error[E0599]: no variant named `Big` found for enum `JoinAllKind<F>`
  --> futures-util/src/future/join_all.rs:45:9
   |
37 | / pin_project_lite::pin_project! {
38 | |     #[project = JoinAllKindProj]
39 | |     pub enum JoinAllKind<F>
40 | |     where
...  |
45 | |         Big  { #[pin] fut: Collect<FuturesOrdered<F>, Vec<F::Output>> },
   | |         ^^^ variant not found in `JoinAllKind<F>`
46 | |     }
47 | | }
   | |_- variant `Big` not found here
   ```
Is this a pin-project-lite problem?

**Comment by taiki-e** [comment]
Yeah, that is https://github.com/taiki-e/pin-project-lite/issues/3#issuecomment-758323290.

File: futures-util/src/future/join_all.rs [hunk 1]
```diff
@@ -19,13 +20,30 @@ fn iter_pin_mut<T>(slice: Pin<&mut [T]>) -> impl Iterator<Item = Pin<&mut T>> {
...
+    #[project = JoinAllKindProj]
+    pub enum JoinAllKind<F>
+    where
+        F: Future,
+    {
+        Small { elems: Pin<Box<[MaybeDone<F>]>> },
+        #[cfg(not(futures_no_atomic_cas))]
+        Big  { #[pin] fut: Collect<FuturesOrdered<F>, Vec<F::Output>> },
+    }
 }
```
Comment (taiki-e) [inline futures-util/src/future/join_all.rs:47]: IIUC, both `Pin<Box<[MaybeDone<F>]>>` and `Collect<FuturesOrdered<F>` are always `Unpin`, so `JoinAllKind` is always `Unpin` and `pin_project!` is not needed here.

**Review by taiki-e — APPROVED** [review]
LGTM. Thanks @ibraheemdev!

**Comment by stuhood** [comment]
Hey folks! Thanks for the optimization. Were any benchmarks performed for this? Also, would the same optimization apply to https://github.com/rust-lang/futures-rs/blob/dd019055ef5bf4309f15db934407e202caf52e14/futures-util/src/future/try_join_all.rs ? It seems to have the same shape.

**Comment by ibraheemdev** [comment]
Good point, `try_join_all` could use the same optimization. Note that this is not a complete fix, if you pass a large iterator without a size hint, the issue is still there. No, no benchmarks were performed, but the issue is quite apparent, although `SMALL` could probably be tweaked.

**Pull request merged**